ray-curator/ray_curator/stages/image/filters/nsfw_filter.py

```diff
@@ -0,0 +1,116 @@
+from dataclasses import dataclass
+
+import numpy as np
+import torch
+from loguru import logger
+
+from ray_curator.backends.base import WorkerMetadata
+from ray_curator.models.nsfw import NSFWScorer
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks import ImageBatch
+
+
+@dataclass
+class ImageNSFWFilterStage(ProcessingStage[ImageBatch, ImageBatch]):
+    """Stage for filtering out NSFW images using NSFWScorer model.
+
+    This class processes image batches through an NSFW scoring model to generate
+    NSFW probability scores for each image. Images with scores above the threshold
+    will be filtered out as NSFW content.
+    """
+    model_dir: str = "models/nsfw"
+    num_gpus_per_worker: float = 0.25
+    model_batch_size: int = 32  # Number of images to process through model at once
+    score_threshold: float = 0.5
+    verbose: bool = False
+
+    @property
+    def name(self) -> str:
+        return "image_nsfw_filter"
+
+    @property
+    def resources(self) -> Resources:
+        return Resources(gpus=self.num_gpus_per_worker)
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def setup(self, worker_metadata: WorkerMetadata | None = None) -> None:  # noqa: ARG002
+        """Initialize the NSFW filtering model."""
+        self.model = NSFWScorer(model_dir=self.model_dir)
+        self.model.setup()
+
+        if self.verbose:
+            logger.info("Initialized NSFW scoring model")
+
+    def process(self, task: ImageBatch) -> ImageBatch:
+        """Process an image batch to generate NSFW scores and filter by threshold.
+
+        Args:
+            task: ImageBatch containing list of ImageObject instances with pre-computed embeddings
+
+        Returns:
+            ImageBatch with filtered images that have NSFW scores below the threshold
+        """
+
+        # Process images in batches to generate scores
+        num_images = len(task.data)
+        for batch_start in range(0, num_images, self.model_batch_size):
```

Comment:
Might be nice to consolidate this into a shared function/class, since it does batching the same way every time. Again will link to how we did it for text: https://github.com/NVIDIA-NeMo/Curator/blob/ray-api/ray-curator/ray_curator/stages/text/models/model.py#L88